homeassistant/components/qbittorrent/sensor.py

```diff
@@ -5,121 +5,191 @@
 import logging
 from typing import Any
+from datetime import datetime, timezone
+
 
 from homeassistant.components.sensor import (
     SensorDeviceClass,
     SensorEntity,
     SensorEntityDescription,
-    SensorStateClass,
 )
 from homeassistant.config_entries import ConfigEntry
 from homeassistant.const import STATE_IDLE, UnitOfDataRate
 from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceEntryType, DeviceInfo
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 from homeassistant.helpers.typing import StateType
 from homeassistant.helpers.update_coordinator import CoordinatorEntity
 
-from .const import DOMAIN
+from .const import (
+    DOMAIN,
+    STATE_DOWNLOADING,
+    STATE_SEEDING,
+    STATE_UP_DOWN,
+)
 from .coordinator import QBittorrentDataCoordinator
 
 _LOGGER = logging.getLogger(__name__)
 
 SENSOR_TYPE_CURRENT_STATUS = "current_status"
 SENSOR_TYPE_DOWNLOAD_SPEED = "download_speed"
 SENSOR_TYPE_UPLOAD_SPEED = "upload_speed"
+SENSOR_TYPE_ALL_TORRENTS = "all_torrents"
+SENSOR_TYPE_PAUSED_TORRENTS = "paused_torrents"
+SENSOR_TYPE_ACTIVE_TORRENTS = "active_torrents"
+SENSOR_TYPE_INACTIVE_TORRENTS = "inactive_torrents"
 
 
-@dataclass(frozen=True)
-class QBittorrentMixin:
-    """Mixin for required keys."""
-
-    value_fn: Callable[[dict[str, Any]], StateType]
-
-
-@dataclass(frozen=True)
-class QBittorrentSensorEntityDescription(SensorEntityDescription, QBittorrentMixin):
-    """Describes QBittorrent sensor entity."""
-
-
-def _get_qbittorrent_state(data: dict[str, Any]) -> str:
-    download = data["server_state"]["dl_info_speed"]
-    upload = data["server_state"]["up_info_speed"]
-
-    if upload > 0 and download > 0:
-        return "up_down"
-    if upload > 0 and download == 0:
-        return "seeding"
-    if upload == 0 and download > 0:
-        return "downloading"
-    return STATE_IDLE
+@dataclass(frozen=True, kw_only=True)
+class QBittorrentSensorEntityDescription(SensorEntityDescription):
+    """Entity description class for qBittorent sensors."""
 
-
-def format_speed(speed):
-    """Return a bytes/s measurement as a human readable string."""
-    kb_spd = float(speed) / 1024
-    return round(kb_spd, 2 if kb_spd < 0.1 else 1)
+    val_func: Callable[[QBittorrentDataCoordinator], StateType]
+    extra_state_attr_func: Callable[[Any], dict[str, str]] | None = None
 
 
 SENSOR_TYPES: tuple[QBittorrentSensorEntityDescription, ...] = (
     QBittorrentSensorEntityDescription(
         key=SENSOR_TYPE_CURRENT_STATUS,
+        translation_key="current_status", 
         name="Status",
-        value_fn=_get_qbittorrent_state,
+        device_class=SensorDeviceClass.ENUM,
+        options=[STATE_IDLE, STATE_UP_DOWN, STATE_SEEDING, STATE_DOWNLOADING],
+        val_func=lambda coordinator: get_state(
+            coordinator.data["server_state"]["up_info_speed"], coordinator.data["server_state"]["dl_info_speed"]
+        ),
```

Comment:
This will probably be the only place where you use the `get_state` function, might as well pass in the full coordinator to make this a bit cleaner